### DISCUSSION TITLE
can I use createContentLoader in config.js?

### DISCUSSION THREAD

**Discussion by Jiny3213** [vuejs/vitepress#2790] [Q&A]
I has a data loader using createContentLoader to load all file in `post/*.md`

Now I want to use it to load my posts in `config.js` to generate the `themeConfig.sidebar`, but the createContentLoader is async load.

How can I do this?

**Comment by brc-dd** [comment]
**EDIT:** won't work for createContentLoader, see comments below

---

async configs are supported. you can have something like this:

```ts
// .vitepress/config.ts

export default async () => {
  // await createContentLoader ...

  return defineConfig({
    // ...
  })
}
```

Haven't checked, but top level awaits are probably also supported.

**Reply by Jiny3213** [reply]
It work, top level awaits are also work!
thanks!!!

**Reply by Jiny3213** [reply]
It seem work when I don't stop the dev server, but clash when I restart the dev server or run build.

error like this:

```
failed to load config from /my/path/.vitepress/config.ts
failed to start server. error:
Error: content loader invoked without an active vitepress process, or before vitepress config is resolved.
```

It seem that createContentLoader can not be call before config is resolved. So I can not use it to  generate the `themeConfig.sidebar` :(

**Comment by Jiny3213** [comment]
Finally, I find a way to override the `VPSidebar.vue` component like https://vitepress.dev/guide/extending-default-theme#overriding-internal-components

``` javascript
// MyVPSidebar.vue
import { data } from '../../posts.data' // it use createContentLoader
const { hasSidebar } = useSidebar()
const sidebarGroups = [{
  items: data.map(item => ({
    text: item.frontmatter.title,
    link: item.url
  }))
}]
// other code...
```
in `.vitepress/config.ts` the themeConfig.sidebar is like:
```
sidebar: [
        {
          text: 'fake sidebar',
          link: '/'
        }
      ]
```
to make `hasSidebar` work

**Comment by astorDev** [comment]
Can we reopen that since @brc-dd answer doesn't help at all with the following problem:

`content loader invoked without an active vitepress process, or before vitepress config is resolved.`

I wasn't able to repeat @Jiny3213 work-around because `MyVPSidebar.vue` code seems incomplete.

**Reply by brc-dd** [reply]
Yeah my answer above seems wrong. It isn't supported outside data loader files. You'll need to manually use glob to workaround that. Feel free to create a feature request though, will see. There are many third party plugins though for auto-generating sidebar which work fine. For example, https://github.com/jooy2/vitepress-sidebar
